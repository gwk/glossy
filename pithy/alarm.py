# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

import signal
from typing import Callable


class Timeout(Exception):
  'Exception indictaing that an AlarmManager timed out.'


class AlarmManager:

  def __init__(self, timeout: int, msg:str='alarm timed out after {timeout} seconds', on_signal: Callable[[], None]=None) -> None:
    if not isinstance(timeout, int): raise TypeError(f'timeout must be an int; received: {timeout!r}')
    if timeout < 0: raise TypeError(f'timeout must be nonnegative; received: {timeout!r}')
    self.timeout = timeout
    self.msg = msg
    self.on_signal = on_signal
    self.timed_out = False


  def __enter__(self):
    if not self.timeout: return
    prev = signal.getsignal(signal.SIGALRM)
    if prev is not signal.SIG_DFL:
      raise Exception(f'task.run encountered previously installed signal handler: {prev}')

    def alarm_handler(signum, current_stack_frame):
      # since signal handlers carry reentrancy concerns; do not do any IO within the handler.
      self.timed_out = True
      if self.on_signal: self.on_signal() # callback must respect reentrancy limitations.

    signal.signal(signal.SIGALRM, alarm_handler) # set handler.
    signal.alarm(self.timeout) # set alarm.


  def __exit__(self, exc_type, exc_value, traceback) -> None:
    if not self.timeout: return None
    signal.alarm(0) # disable alarm.
    signal.signal(signal.SIGALRM, signal.SIG_DFL) # uninstall handler.
    if exc_type: return None # exception will be reraised.
    if self.timed_out: raise Timeout(self.msg.format(timeout=self.timeout))
    return None
