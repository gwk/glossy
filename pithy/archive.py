# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from functools import total_ordering
from io import BytesIO, TextIOBase, TextIOWrapper
from typing import Any, BinaryIO, Callable, Iterable, List, Tuple, Union
from tarfile import TarFile
from zipfile import ZipFile, BadZipFile
from .fs import path_ext
from .util import lazy_property
from .io import *


FileOrPath = Union[BinaryIO, str] # TODO: support other pathlike objects.
Opener = Callable[[], BinaryIO]


class Archive:

  def __init__(self, file_or_path:FileOrPath, ext=None) -> None:
    self.name: str = getattr(file_or_path, 'name', str(file_or_path))
    if ext is None:
      ext = path_ext(self.name)
    self.ext = ext
    try: self._handler = _ext_handlers[ext](file_or_path)
    except KeyError: raise ValueError(f'Archive does not support extension: {ext!r}')


  def __repr__(self) -> str:
    return f'{self.__class__.__name__}(name={self.name!r})'


  def __iter__(self) -> Iterable['ArchiveFile']:
    return self._handler.files(self)



@total_ordering
class ArchiveMember:

  def __init__(self, archive: Archive, name: str) -> None:
    self.archive = archive
    self.name = name

  def __repr__(self) -> str:
    return f'{self.__class__.__name__}({self.archive}, {self.name!r})'

  def __lt__(self, other: Any) -> bool:
    if isinstance(other, ArchiveMember):
      return self.name < other.name
    return NotImplemented



class ArchiveFile(ArchiveMember):

  def __init__(self, archive:Archive, name:str, opener:Opener) -> None:
    super().__init__(archive=archive, name=name)
    self._opener = opener

  @lazy_property
  def _file(self): return self._opener()

  def __iter__(self): return iter(self._file)

  def __getattr__(self, name):
    # note: dunder attribute access is never forwarded to this method.
    if name.startswith('_'): raise AttributeError(name)
    return getattr(self._file, name)

  def seekable(self):
    return getattr(self._file, 'seekable', False)

  def text(self, encoding='UTF-8', errors=None, newline=None) -> TextIOWrapper:
    'Returns a TextIOWrapper around the underlying binary buffer.'
    return TextIOWrapper(self._file, encoding=encoding, errors=errors, newline=newline)



class ArchiveDir(ArchiveMember): pass


class _Handler:
  def __init__(self, file_or_path:FileOrPath) -> None: raise NotImplementedError()


class _TarHandler(_Handler):
  def __init__(self, file_or_path:FileOrPath) -> None:
    self.tar = TarFile.open(mode='r|*', fileobj=file_or_path)

  def files(self, archive:Archive) -> Iterable[Tuple[str, Opener]]:
    for info in self.tar:
      if info.isdir(): continue
      opener = lambda: self.tar.extractfile(info)
      #^ note: crucial to pass info here and not name, or else gz archives get very slow.
      yield ArchiveFile(archive=archive, name=info.name, opener=opener)


class _ZipHandler(_Handler):
  def __init__(self, file_or_path:FileOrPath) -> None:
    try: self.zip = ZipFile(file_or_path)
    except BadZipFile as e:
      if isinstance(file_or_path, TextIOBase):
        raise TypeError(f'Archive requires a path or binary stream but received text stream: {file_or_path}') from e
      raise

  def files(self, archive:Archive) -> Iterable[Tuple[str, Opener]]:
    for name in self.zip.namelist():
      if name.endswith('/'): continue # directory.
      opener = lambda: self.zip.open(name)
      yield ArchiveFile(archive=archive, name=name, opener=opener)


_ext_handlers = {
  '.bz2' : _TarHandler,
  '.gz'  : _TarHandler,
  '.tar' : _TarHandler,
  '.xz', : _TarHandler,
  '.zip' : _ZipHandler,
}
