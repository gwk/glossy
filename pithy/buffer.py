# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from typing import Callable, Generic, Iterable, Iterator, List, TypeVar, Union
from .default import Raise, RaiseOr


T = TypeVar('T')


class Buffer(Iterator[T]):
  '''
  Iterable object that buffers an iterable.
  Call push() to push an item into the buffer;
  this will be returned on the subsequent call to next().
  '''


  def __init__(self, iterable: Iterable[T]) -> None:
    self.iterator = iter(iterable)
    self.buffer: List[T] = []


  def __repr__(self) -> str:
    return 'IterBuffer({!r}, buffer={!r})'.format(self.iterator, self.buffer)


  def __iter__(self) -> Iterator[T]: return self


  def __next__(self) -> T:
    try: return self.buffer.pop()
    except IndexError: pass
    return next(self.iterator)


  @property
  def is_live(self) -> bool:
    try: self.peek()
    except StopIteration: return False
    else: return True


  def push(self, item: T) -> None:
    self.buffer.append(item)


  def peek(self, default: RaiseOr[T]=Raise._) -> T:
    try: return self.buffer[-1]
    except IndexError: pass
    try: el = next(self.iterator)
    except StopIteration:
      if isinstance(default, Raise): raise
      else: return default
    self.buffer.append(el)
    return el


  def take_while(self, pred: Callable[[T], bool]) -> Iterator[T]:
    for el in self:
      if pred(el):
        yield el
      else:
        self.buffer.append(el)
        break


  def drop_while(self, pred: Callable[[T], bool]) -> None:
    for el in self:
      if not pred(el):
        self.buffer.append(el)
        break


  def peek_while(self, pred: Callable[[T], bool]) -> List[T]:
    els = list(self.take_while(pred))
    self.buffer.extend(reversed(els))
    return els


  def take(self, count: int, short=False, default: RaiseOr[T]=Raise._) -> List[T]:
    els = []
    for _ in range(count):
      try: els.append(next(self))
      except StopIteration:
        if short: break
        if isinstance(default, Raise): raise
        els.append(default)
    return els


  def peeks(self, count: int, short=False, default: RaiseOr[T]=Raise._) -> List[T]:
    if 0 < count <= len(self.buffer):
      return list(reversed(self.buffer[-count:]))
    els = []
    for _ in range(count):
      try: els.append(next(self))
      except StopIteration:
        if short: break
        if isinstance(default, Raise): raise
        els.append(default)
    self.buffer.extend(reversed(els))
    return els


  def expect(self, pred: Callable[[T], bool]) -> T:
    el = next(self)
    if pred(el): return el
    raise ValueError(el)
