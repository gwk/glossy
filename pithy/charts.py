from .svg import *
from io import StringIO
from typing import *

class XYChart:
  def __init__(self, *series:Tuple[str, Iterable[Tuple]],
   title:str=None,
   x_label:str=None, y_label:str=None,
   w:int=None, h:int=None,
   plot_w:int=None, plot_h:int=None,
   x_min:int=None, x_max:int=None, y_min:int=None, y_max:int=None,
   title_height:int=None,
   label_width:int=None, label_height:int=None,
   legend_width:int=None, # TODO: legend_height would imply that the legend is at bottom.
   ) -> None:
    self.series = {k : list(v) for k, v in series}

    #take smallest_x and largest_x value from data if x_min or y_min is undefined
    min_x:float = 0
    max_x:float = 1
    min_y:float = 0
    max_y:float = 1
    for k,s in self.series.items():
      for x,y in s:
        min_x = min(x, min_x)
        max_x = max(x, max_x)
        min_y = min(y, min_y)
        max_y = max(y, max_y)
    if(x_min == None): x_min = min_x
    if(y_min == None): y_min = min_y
    if(x_max == None): x_max = max_x
    if(y_max == None): y_max = max_y
    self.x_min = x_min
    self.x_max = x_max
    self.y_min = y_min
    self.y_max = y_max
    x_range:float = x_max - x_min
    y_range:float = y_max - y_min
    self.svg_width = w
    self.svg_height = h
    self.label_height = label_height
    self.label_width = label_width
    self.legend_width = legend_width
    self.title_height = title_height
    self.title = title
    self.x_label = x_label
    self.y_label = y_label

    x_scale = ((self.svg_width - 10) - (self.legend_width + self.label_width + 10))/x_range
    y_scale = ((self.svg_height - self.title_height)-(self.label_height+10))/y_range
    scaled_min_x=0
    scaled_min_y=0
    for k,s in self.series.items():
      for x,y in s:
        scaled_min_x = min(x*x_scale, scaled_min_x)
        scaled_min_y = min(y*y_scale, scaled_min_y)
    for k,s in self.series.items():
      temp_list = []
      for x,y in s:
        temp_list.append(((x*x_scale)-scaled_min_x, (y*y_scale)-scaled_min_y))
      self.series[k] = temp_list



  def render(self, file=None) -> str:
    colors:List = ['#EF7F6D', '#F78398', '#E893C4', '#C2AAE5', '#89C2F3', '#49D5E9', '#38E2CA', '#70E9A0', '#AFEB77', '#EFE55D']
    if file is None:
      file = StringIO()
    chart_x_min:int = self.legend_width + self.label_width + 10
    chart_x_max:int = self.svg_width - 10
    chart_y_min:int = self.label_height+10
    chart_y_max:int = self.svg_height - self.title_height
    with SvgWriter(file, w=self.svg_width, h=self.svg_height) as svg:
      y_range = chart_y_max - chart_y_min
      #draw graph
      svg.rect(pos=(chart_x_min, chart_y_min), size=(chart_x_max-chart_x_min, chart_y_max - chart_y_min), stroke=None, fill='#fafafa')
      transform_x =  700
      color_picker = 0
      with svg.g(transform=f'translate({chart_x_min} {chart_y_max}) scale(1 -1) '):
        for i in range(1,5):
            svg.line((0, (y_range/5)*i), (chart_x_max-70, (y_range/5)*i), stroke='#BBCCDD' )
            with svg.g(transform=' scale(1 -1) '):
              svg.text(text=((self.y_max-self.y_min)/5*i), x = -40, y = -((y_range/5)*i) + 6, style="color:black") #max y
        for k,s in self.series.items(): #each series
          for p in s:
            print(p)
          svg.polyline((p for p in s), style=f'fill:none;stroke:{colors[color_picker]};stroke-width:1')
          color_picker = (1+color_picker) % 10
          print(color_picker)

      #text
      svg.text(text=self.title, x=self.svg_width/1.9, y = self.label_height*1.4, style=f"color:black;font-size:{self.title_height}") #title
      svg.text(text=self.y_label, x=0, y = self.svg_height/2, style="color:black") #x axis
      svg.text(text=self.x_label, x=self.svg_width/2, y = self.svg_height, style="color:black") #y axis
      svg.text(text=str(self.x_min), x = chart_x_min, y = (chart_y_max+15), style="color:black") #min x
      svg.text(text=str(self.y_min), x = chart_x_min-10, y = (chart_y_max), style="color:black") #min y
      svg.text(text=str(self.x_max), x = chart_x_max-20, y = (chart_y_max+15), style="color:black") #max x
      svg.text(text=str(self.y_max), x = chart_x_min-25, y = (chart_y_min+10), style="color:black") #max y
      #horizontal lines




      pass
