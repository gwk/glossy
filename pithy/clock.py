# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from sys import stderr
from time import time as now


class Clock:
  def __init__(self, name:str):
    self.name = name
    self.times = [(now(), 'start')]

  def split(self, label:str):
    t = now()
    print(f'clock: {self.name}: {label}: {t - self.times[-1][0]:.04f}', file=stderr)
    self.times.append((t, label))


  def total(self):
    t = now()
    print(f'clock: {self.name}: TOTAL: {t - self.times[0][0]:.04f}', file=stderr)
