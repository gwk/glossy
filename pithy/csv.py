# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

# TODO: handle leading UTF8-BOM.

import csv
import collections
from sys import stdout
from types import TracebackType
from typing import ContextManager, Iterable, Iterator, Optional, Sequence, TextIO, Type, Union


def load_csv(file: TextIO,
 dialect:Optional[str]=None,
 delimiter:Optional[str]=None,
 doublequote:Optional[bool]=None,
 escapechar:Optional[str]=None,
 quotechar:Optional[str]=None,
 quoting:Optional[int]=None,
 skipinitialspace:Optional[bool]=None,
 strict:Optional[bool]=None,
 header:Union[None, bool, Sequence[str]]=None) -> 'CSVFileReader':

  return CSVFileReader(
    file=file,
    dialect=dialect,
    delimiter=delimiter,
    doublequote=doublequote,
    escapechar=escapechar,
    quotechar=quotechar,
    quoting=quoting,
    skipinitialspace=skipinitialspace,
    strict=strict,
    header=header)


def write_csv(f:TextIO, *, header:Optional[Sequence[str]], rows:Iterable[Sequence]) -> None:
  w = csv.writer(f)
  if header is not None: w.writerow(header)
  w.writerows(rows)


def out_csv(*, header:Optional[Sequence[str]], rows:Iterable[Sequence]) -> None:
  write_csv(f=stdout, header=header, rows=rows)


class CSVFileReader(Iterable, ContextManager):

  def __init__(self, file: TextIO,
   dialect:Optional[str]=None,
   delimiter:Optional[str]=None,
   doublequote:Optional[bool]=None,
   escapechar:Optional[str]=None,
   quotechar:Optional[str]=None,
   quoting:Optional[int]=None,
   skipinitialspace:Optional[bool]=None,
   strict:Optional[bool]=None,
   header:Union[None, bool, Sequence[str]]=None) -> None:

    opts = { k : v for (k, v) in [
      ('dialect', dialect),
      ('delimiter', delimiter),
      ('doublequote', doublequote),
      ('escapechar', escapechar),
      ('quotechar', quotechar),
      ('quoting', quoting),
      ('skipinitialspace', skipinitialspace),
      ('strict', strict),
      ] if v is not None }

    self._reader = csv.reader(file, dialect, **opts)
    self.file = file

    if header is None or isinstance(header, bool):
      if header: next(self._reader) # simply discard.
    else: # match expected against actual.
      row = next(self._reader)
      list_header = list(header)
      if row != list_header:
        raise ValueError(f'load_csv expected header:\n{list_header}\nreceived:\n{row}')


  def __iter__(self) -> Iterator[Sequence[str]]:
    return self._reader

  def __enter__(self) -> 'CSVFileReader':
    return self

  def __exit__(self, exc_type: Optional[Type[BaseException]], exc_value: Optional[BaseException],
   traceback: Optional[TracebackType]) -> None:
    self.file.close()
