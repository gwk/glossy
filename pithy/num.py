# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

'''
Numeric utilities.
'''

from .dataclasses import dataclass
from typing import Sequence, Iterator, Union


Num = Union[int, float]


_set_attr = object.__setattr__


@dataclass(frozen=True)
class NumRange(Sequence[Num]):

  start:Num
  stop:Num
  step:Num
  _len:int

  def __init__(self, start:Num, stop:Num=None, step:Num=1) -> None:
    if stop is None: # Imitate `range`; `start` is actually `stop`.
      _set_attr(self, 'start', 0)
      _set_attr(self, 'stop', start)
    else:
      _set_attr(self, 'start', start)
      _set_attr(self, 'stop', stop)
    _set_attr(self, 'step', step)
    dist = max(0, self.stop - self.start)
    steps = dist / step
    extra = 1 if steps % 1 else 0
    _set_attr(self, '_len', int(steps) + extra)

  def __len__(self):
    return self._len

  def __getitem__(self, i:Union[int, slice]) -> Num: # type: ignore
    if isinstance(i, int):
      if i >= 0 and i >= self._len: raise IndexError(i)
      return self.start + self.step * (i if i >= 0 else (self._len+i))
    elif isinstance(i, slice):
      raise NotImplementedError('NumRange does not yet support extended slicing')

  def __iter__(self) -> Iterator[Num]:
    return (self.start + self.step * i for i in range(self._len))

  def __repr__(self):
    return '{}({}, {}, {})'.format(type(self).__name__, self.start, self.stop, self.step)
