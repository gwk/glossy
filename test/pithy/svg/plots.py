#!/usr/bin/env python3

from pithy.svg import *
from pithy.num import NumRange
from sys import stdout
from math import sin, pi


size = (400,200)

with Svg(stdout, w=1000, h=1000, stroke='black', fill='white', stroke_width=1) as svg:

  y = 5
  r = NumRange(0, 1, 1/16, closed=True)
  with svg.plot(pos=(5, y), size=size, title_h=12, tick_w=32, tick_h=12,
   title='Identity and Sin',
   grid_step=0.25,
   tick_step=(0.25, 0.25),
   tick_x=True, tick_y=True,
   series=[
    LineSeries('identity', [(x, x) for x in r]),
    LineSeries('sin', [(x, sin(x*2*pi)) for x in r])
   ]):
    pass
