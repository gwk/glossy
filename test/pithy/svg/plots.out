<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" stroke="black" fill="white" stroke-width="1" width="1000" height="1000">
<g transform="translate(5,5)" class="plot">
<style>
text {
  stroke: none;
  fill: currentColor;
}
text.title {
  text-anchor: start;
  alignment-baseline: hanging;
}
g.tick-x text.tick {
  white-space: pre;
  text-anchor: start;
  alignment-baseline: hanging;
}
g.tick-y text.tick {
  white-space: pre;
  text-anchor: start;
  alignment-baseline: alphabetic;
}
g.legend rect {
  fill: none;
}
g.legend text {
  text-anchor: start;
  alignment-baseline: middle;
}
text.title { font-size: 14.0px; }
text.axis-label { font-size: 12.0px; }
text.tick { font-size: 10.0px; }
tspan.zpad { opacity: 0; }
</style>
<text class="title" x="0" y="0">Identity and Sin</text>
<g transform="translate(0,14)">
<clipPath id="_id0">
<rect width="378" height="146"/>
</clipPath>
<g class="grid">
<line x1="18.9" y1="0" x2="18.9" y2="146"/>
<line x1="37.8" y1="0" x2="37.8" y2="146"/>
<line x1="56.7" y1="0" x2="56.7" y2="146"/>
<line x1="75.6" y1="0" x2="75.6" y2="146"/>
<line x1="94.5" y1="0" x2="94.5" y2="146"/>
<line x1="113.4" y1="0" x2="113.4" y2="146"/>
<line x1="132.3" y1="0" x2="132.3" y2="146"/>
<line x1="151.2" y1="0" x2="151.2" y2="146"/>
<line x1="170.1" y1="0" x2="170.1" y2="146"/>
<line x1="189" y1="0" x2="189" y2="146"/>
<line x1="207.9" y1="0" x2="207.9" y2="146"/>
<line x1="226.8" y1="0" x2="226.8" y2="146"/>
<line x1="245.7" y1="0" x2="245.7" y2="146"/>
<line x1="264.6" y1="0" x2="264.6" y2="146"/>
<line x1="283.5" y1="0" x2="283.5" y2="146"/>
<line x1="302.4" y1="0" x2="302.4" y2="146"/>
<line x1="321.3" y1="0" x2="321.3" y2="146"/>
<line x1="340.2" y1="0" x2="340.2" y2="146"/>
<line x1="359.1" y1="0" x2="359.1" y2="146"/>
<line x1="0" y1="109.5" x2="378" y2="109.5"/>
<line x1="0" y1="73" x2="378" y2="73"/>
<line x1="0" y1="36.5" x2="378" y2="36.5"/>
<rect class="grid-border" fill="none" x="0" y="0" width="378" height="146"/>
</g>
<line class="axis" id="x-axis" x1="0" y1="73" x2="378" y2="73"/>
<line class="axis" id="y-axis" x1="0" y1="0" x2="0" y2="146"/>
<g class="tick-x">
<line class="tick" x1="0" y1="146" x2="0" y2="150"/>
<text class="tick" x="0" y="151"><tspan class="zpad"></tspan>0.0</text>
<line class="tick" x1="37.8" y1="146" x2="37.8" y2="150"/>
<text class="tick" x="37.8" y="151"><tspan class="zpad"></tspan>0.1</text>
<line class="tick" x1="75.6" y1="146" x2="75.6" y2="150"/>
<text class="tick" x="75.6" y="151"><tspan class="zpad"></tspan>0.2</text>
<line class="tick" x1="113.4" y1="146" x2="113.4" y2="150"/>
<text class="tick" x="113.4" y="151"><tspan class="zpad"></tspan>0.3</text>
<line class="tick" x1="151.2" y1="146" x2="151.2" y2="150"/>
<text class="tick" x="151.2" y="151"><tspan class="zpad"></tspan>0.4</text>
<line class="tick" x1="189" y1="146" x2="189" y2="150"/>
<text class="tick" x="189" y="151"><tspan class="zpad"></tspan>0.5</text>
<line class="tick" x1="226.8" y1="146" x2="226.8" y2="150"/>
<text class="tick" x="226.8" y="151"><tspan class="zpad"></tspan>0.6</text>
<line class="tick" x1="264.6" y1="146" x2="264.6" y2="150"/>
<text class="tick" x="264.6" y="151"><tspan class="zpad"></tspan>0.7</text>
<line class="tick" x1="302.4" y1="146" x2="302.4" y2="150"/>
<text class="tick" x="302.4" y="151"><tspan class="zpad"></tspan>0.8</text>
<line class="tick" x1="340.2" y1="146" x2="340.2" y2="150"/>
<text class="tick" x="340.2" y="151"><tspan class="zpad"></tspan>0.9</text>
<line class="tick" x1="378" y1="146" x2="378" y2="150"/>
<text class="tick" x="378" y="151"><tspan class="zpad"></tspan>1.0</text>
</g>
<g class="tick-y">
<line class="tick" x1="378" y1="146" x2="382" y2="146"/>
<text class="tick" x="383" y="146"><tspan class="zpad"></tspan>-1.0</text>
<line class="tick" x1="378" y1="109.5" x2="382" y2="109.5"/>
<text class="tick" x="383" y="109.5"><tspan class="zpad"></tspan>-0.5</text>
<line class="tick" x1="378" y1="73" x2="382" y2="73"/>
<text class="tick" x="383" y="73"><tspan class="zpad">0</tspan>0.0</text>
<line class="tick" x1="378" y1="36.5" x2="382" y2="36.5"/>
<text class="tick" x="383" y="36.5"><tspan class="zpad">0</tspan>0.5</text>
<line class="tick" x1="378" y1="0" x2="382" y2="0"/>
<text class="tick" x="383" y="0"><tspan class="zpad">0</tspan>1.0</text>
</g>
<g class="legend">
<g id="legend-identity">
<rect x="1" y="169" width="16" height="16"/>
<text x="21" y="177">identity</text>
</g>
<g id="legend-sin">
<rect x="129" y="169" width="16" height="16"/>
<text x="149" y="177">sin</text>
</g>
</g>
<g class="series" clip-path="url(#_id0)">
<g id="identity">
<polyline fill="none" points="0,73 23.6,68.4 47.2,63.9 70.9,59.3 94.5,54.8 118.1,50.2 141.8,45.6 165.4,41.1 189,36.5 212.6,31.9 236.2,27.4 259.9,22.8 283.5,18.2 307.1,13.7 330.8,9.1 354.4,4.6 378,0"/>
</g>
<g id="sin">
<polyline fill="none" points="0,73 23.6,45.1 47.2,21.4 70.9,5.6 94.5,0 118.1,5.6 141.8,21.4 165.4,45.1 189,73 212.6,100.9 236.2,124.6 259.9,140.4 283.5,146 307.1,140.4 330.8,124.6 354.4,100.9 378,73"/>
</g>
</g>
</g>
</g>
</svg>
